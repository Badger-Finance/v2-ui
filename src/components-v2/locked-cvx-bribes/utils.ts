import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
