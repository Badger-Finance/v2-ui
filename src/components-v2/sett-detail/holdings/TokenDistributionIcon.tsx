import React from 'react';
import { makeStyles, Tooltip } from '@material-ui/core';
import { TokenDistribution } from './TokenDistribution';
import HelpIcon from '@material-ui/icons/Help';
import { SettData } from '@badger-dao/sdk';

const useStyles = makeStyles((theme) => ({
	helpIcon: {
		fontSize: 16,
		marginLeft: theme.spacing(0.5),
		color: 'rgba(255, 255, 255, 0.3)',
	},
}));

interface Props {
	settBalance: SettData;
}

export const TokenDistributionIcon = ({ settBalance }: Props): JSX.Element | null => {
	const classes = useStyles();

	if (settBalance.tokens.length === 0) {
		return null;
	}

	return (
		<Tooltip
			// not adding an id makes mui generate a random id on each render which breaks the tests snapshots
			// see https://github.com/mui-org/material-ui/issues/21293
			id="sett-token-distribution-tooltip"
			aria-label="sett token distribution"
			enterTouchDelay={0}
			arrow
			placement="top"
			title={<TokenDistribution settBalance={settBalance} />}
		>
			<HelpIcon className={classes.helpIcon} />
		</Tooltip>
	);
};
