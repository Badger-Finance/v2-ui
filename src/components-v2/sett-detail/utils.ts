import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
