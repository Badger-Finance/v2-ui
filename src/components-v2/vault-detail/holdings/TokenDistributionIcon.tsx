import { VaultData } from '@badger-dao/sdk';
import { makeStyles, Tooltip } from '@material-ui/core';
import HelpIcon from '@material-ui/icons/Help';
import React from 'react';

import { TokenDistribution } from './TokenDistribution';

const useStyles = makeStyles((theme) => ({
  helpIcon: {
    fontSize: 16,
    marginLeft: theme.spacing(0.5),
    color: 'rgba(255, 255, 255, 0.3)',
  },
}));

interface Props {
  settBalance: VaultData;
}

export const TokenDistributionIcon = ({ settBalance }: Props): JSX.Element | null => {
  const classes = useStyles();

  if (settBalance.tokens.length === 0) {
    return null;
  }

  return (
    <Tooltip
      // not adding an id makes mui generate a random id on each render which breaks the tests snapshots
      // see https://github.com/mui-org/material-ui/issues/21293
      id="sett-token-distribution-tooltip"
      aria-label="sett token distribution"
      enterTouchDelay={0}
      arrow
      placement="top"
      title={<TokenDistribution settBalance={settBalance} />}
    >
      <HelpIcon className={classes.helpIcon} />
    </Tooltip>
  );
};
