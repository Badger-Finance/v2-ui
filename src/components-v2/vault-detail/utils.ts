import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
