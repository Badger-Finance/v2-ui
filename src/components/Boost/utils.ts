import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
