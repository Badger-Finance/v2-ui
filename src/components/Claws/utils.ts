import BigNumber from 'bignumber.js';
import { INVALID_REASON } from './index';

export interface ValidationParams {
	amount: string | BigNumber;
	maximum: BigNumber;
	minimum?: BigNumber;
}

export function validateAmountBoundaries({
	amount,
	maximum,
	minimum = new BigNumber('0'),
}: ValidationParams): INVALID_REASON | undefined {
	let error: INVALID_REASON | undefined = undefined;

	const input = new BigNumber(amount);
	const amountExceedsBalance = input.gt(maximum);
	const amountIsLessThanMinimum = input.lt(minimum);

	if (amountIsLessThanMinimum) error = INVALID_REASON.UNDER_MINIMUM;
	if (amountExceedsBalance) error = INVALID_REASON.OVER_MAXIMUM;

	console.log({
		input: input.toString(),
		minimum: minimum.toString(),
		maximum: maximum.toString(),
		amountExceedsBalance,
		amountIsLessThanMinimum,
		error,
	});

	return error;
}
