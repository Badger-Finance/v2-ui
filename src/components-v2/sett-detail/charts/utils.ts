import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
