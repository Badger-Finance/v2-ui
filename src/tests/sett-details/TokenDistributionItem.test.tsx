import React from 'react';
import { SAMPLE_SETT_BALANCE } from '../utils/samples';
import { TokenDistributionIcon } from '../../components-v2/sett-detail/holdings/TokenDistributionIcon';
import { customRender, screen } from '../Utils';
import userEvent from '@testing-library/user-event';
import { StoreProvider } from '../../mobx/store-context';
import store from '../../mobx/RootStore';

describe('Token Distribution Icon', () => {
  it('displays sett token distribution information on hover', async () => {
    const { baseElement } = customRender(
      <StoreProvider value={store}>{<TokenDistributionIcon settBalance={SAMPLE_SETT_BALANCE} />}</StoreProvider>,
    );
    userEvent.hover(screen.getByLabelText('sett token distribution'));
    await screen.findByText(SAMPLE_SETT_BALANCE.tokens[0].symbol);
    expect(baseElement).toMatchSnapshot();
  });
});
