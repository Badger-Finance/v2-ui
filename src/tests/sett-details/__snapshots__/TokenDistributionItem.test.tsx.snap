// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token Distribution Icon displays sett token distribution information on hover 1`] = `
<body>
   
   
  
  
  <div>
    <svg
      aria-describedby="mui-31526"
      aria-hidden="true"
      aria-label="sett token distribution"
      class="MuiSvgIcon-root-13 makeStyles-helpIcon-1"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
      />
    </svg>
  </div>
  <div
    class="MuiTooltip-popper-2 MuiTooltip-popperArrow-4"
    id="mui-31526"
    role="tooltip"
    style="position: absolute; will-change: transform; top: 0px; left: 0px; transform: translate3d(NaNpx, NaNpx, 0);"
    x-placement="top"
  >
    <div
      class="MuiTooltip-tooltip-5 MuiTooltip-tooltipPlacementTop-11 MuiTooltip-tooltipArrow-6"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 133ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiGrid-root-25 makeStyles-root-22 MuiGrid-container-26"
      >
        <p
          class="MuiTypography-root-128 makeStyles-title-23 MuiTypography-body1-130"
        >
          Token Distribution
        </p>
        <div
          class="MuiGrid-root-25 makeStyles-content-24 MuiGrid-container-26"
        >
          <div
            class="MuiGrid-root-25 MuiGrid-container-26 MuiGrid-align-items-xs-center-34 MuiGrid-justify-xs-space-between-45"
          >
            <div
              class="makeStyles-tokenNameContainer-158"
            >
              <div
                class="makeStyles-tokenIconContainer-160"
              >
                <img
                  alt="token distribution item"
                  class="makeStyles-icon-161"
                  src="/assets/icons/renbtc.png"
                />
              </div>
              <p
                class="MuiTypography-root-128 makeStyles-tokenInfo-159 MuiTypography-body1-130 MuiTypography-displayInline-156"
              >
                renBTC
              </p>
            </div>
            <p
              class="MuiTypography-root-128 makeStyles-tokenInfo-159 MuiTypography-body1-130 MuiTypography-displayInline-156"
            >
              478.88017951
            </p>
          </div>
          <div
            class="MuiGrid-root-25 MuiGrid-container-26 MuiGrid-align-items-xs-center-34 MuiGrid-justify-xs-space-between-45"
          >
            <div
              class="makeStyles-tokenNameContainer-158"
            >
              <div
                class="makeStyles-tokenIconContainer-160"
              >
                <img
                  alt="token distribution item"
                  class="makeStyles-icon-161"
                  src="/assets/icons/wbtc.png"
                />
              </div>
              <p
                class="MuiTypography-root-128 makeStyles-tokenInfo-159 MuiTypography-body1-130 MuiTypography-displayInline-156"
              >
                WBTC
              </p>
            </div>
            <p
              class="MuiTypography-root-128 makeStyles-tokenInfo-159 MuiTypography-body1-130 MuiTypography-displayInline-156"
            >
              472.09728496
            </p>
          </div>
          <div
            class="MuiGrid-root-25 MuiGrid-container-26 MuiGrid-align-items-xs-center-34 MuiGrid-justify-xs-space-between-45"
          >
            <div
              class="makeStyles-tokenNameContainer-158"
            >
              <div
                class="makeStyles-tokenIconContainer-160"
              >
                <img
                  alt="token distribution item"
                  class="makeStyles-icon-161"
                  src="/assets/icons/sbtc.png"
                />
              </div>
              <p
                class="MuiTypography-root-128 makeStyles-tokenInfo-159 MuiTypography-body1-130 MuiTypography-displayInline-156"
              >
                sBTC
              </p>
            </div>
            <p
              class="MuiTypography-root-128 makeStyles-tokenInfo-159 MuiTypography-body1-130 MuiTypography-displayInline-156"
            >
              337.6574109700648
            </p>
          </div>
        </div>
      </div>
      <span
        class="MuiTooltip-arrow-7"
      />
    </div>
  </div>
</body>
`;
